lib/logstash/filters/grok/timeout_enforcer.rb

```diff
@@ -64,44 +61,26 @@ def stop!
 
   def start_thread_groking(thread)
     # Clear any interrupts from any previous invocations that were not caught by Joni
-    thread.interrupted
-    synchronize do
-      @threads_to_start_time[thread] = java.lang.System.nanoTime()
-    end
+    java.lang.Thread.interrupted
+    @threads_to_start_time.put(thread, java.lang.System.nanoTime)
   end
 
   def stop_thread_groking(thread)
-    synchronize do
-      @threads_to_start_time.delete(thread)
-    end
+    @threads_to_start_time.delete(thread)
   end
 
   def cancel_timed_out!
-    synchronize do
-      @threads_to_start_time.each do |thread,start_time|
-        now = java.lang.System.nanoTime # save ourselves some nanotime calls
-        elapsed = java.lang.System.nanoTime - start_time
-        if elapsed > @timeout_nanos
-          elapsed_millis = elapsed / 1000
-          thread.interrupt()
-          # Ensure that we never attempt to cancel this thread twice in the event
-          # of weird races
-          stop_thread_groking(thread)
-        end
+    now = java.lang.System.nanoTime # save ourselves some nanotime calls
+    @threads_to_start_time.entry_set.each do |entry|
```

Comment:
Isn't this a race condition?

The following sequence seems possible:

1. Thread1 starts grokking
2. Thread 1 exceeds timeout
3. Timeout Enforcer notices that Thread1 exceeds timeout
4. Thread 1 stops grokking
5. Thread 1 Starts a new grokking
6. Thread 1 is now interrupted

It seems to me that we would need to use `ConcurrentHashMap.forEach`.

Here's what the docs say:

```
ConcurrentHashMaps support a set of sequential and parallel bulk operations that, unlike most Stream methods, are designed to be safely, and often sensibly, applied even with maps that are being concurrently updated by other threads; for example, when computing a snapshot summary of the values in a shared registry. There are three kinds of operation, each with four forms, accepting functions with Keys, Values, Entries, and (Key, Value) arguments and/or return values. Because the elements of a ConcurrentHashMap are not ordered in any particular way, and may be processed in different orders in different parallel executions, the correctness of supplied functions should not depend on any ordering, or on any other objects or values that may transiently change while computation is in progress; and except for forEach actions, should ideally be side-effect-free. Bulk operations on Map.Entry objects do not support method setValue.
```

From: https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#forEachKey-long-java.util.function.Consumer-

WDYT @original-brownbear ?

Reply:
@andrewvc that's why I changed the `if` to also check for `start_time < now` to catch this case :)

Reply:
There is still a race however after that point and before the interrupt right?

I think that's unlikely to happen, but with enough iterations...

Reply:
@andrewvc uff, you're right. Sorry for missing this + thanks for catching this !:)

Give me a few minutes to implement a guard against this without locking :)

Reply:
One thought, we only need to lock a thread (which maps to one hash key) at a time.

It may be less lock-y to grab all the keys, then run ConcurrentHashMap.compute on each key. I'm not sure what the granularity of that lock is, but I'd hope that it's per key.

It's more locking / unlocking, but it might be more granular and block less.

Reply:
Another thought, does the pipeline enforcer locking really hurt much? it should execute relatively quickly, we should be able to reduce the cost of locking by increasing the amount of time between cleanings.

Reply:
@andrewvc sadly the locking hurts quite a bit even when uncontended :(

I think there is a clean and very fast solution here, by simply using modification counting.
I actually didn't get around to this today but will in all likelihood implement it tomorrow on top of this.
In a sense all, you need to ensure before you cancel a thread is that the number of events/fields that passed through it is the same as it was when you compared the elapsed time.

So you can probably just have a counter per worker and atomic(or just volatile)-boolean guarding updates to the counter and that workers busy wait on (that's fine because it happens so rarely and so quickly that you spin for fewer cycles than an actual `lock` would take).

=> Single writer principle in action, boolean is only written to by the timeout enforcer, counters are only written to by the worker threads
=> lock-free and probably almost no performance hit compared what is in this PR since all workers will be hit with mostly just reading one `volatile` `boolean` and writing one `volatile` `int` per call and the timeout enforcer will be hit by writing the boolean and reading the ints.
... Disruptor light so to say :)
=> I'm pretty confident we can do this lock-free without a big change :)

Reply:
I see that the locking issues are a bit tricky. If you pardon the pun does this mean this PR is    now deadlocked? 😉

Reply:
@IrlJidel I think we have other options as well :)

@original-brownbear what about inverting the threading as we discussed before and having 2 threads per grok, one tracking the timeout, the other doing the grokking and having `Thread.interrupt` called on it?

Reply:
@IrlJidel don't worry I'll get back to this shortly :)

@andrewvc I think using mod_count is easier actually. Was a little busy with other stuff, so I haven't implemented it yet, but it'll probably be like 30 lines tops to do this here and we don't have to do a huge risky change.
Let's go for that imo, many threads + JRuby isn't a winning combination due to the high overhead per thread memory wise.

Reply:
@andrewvc actually come to think of it it's a 5 line solution ... will fix this a little later today :)